kv/kvserver: TestMergeQueueDoesNotInterruptReplicationChange failed

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9214232?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9214232?buildTab=artifacts#/) on master @ [b1f889e669d435dfff4118e34e98d84af307d888](https://github.com/cockroachdb/cockroach/commits/b1f889e669d435dfff4118e34e98d84af307d888):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2069415939
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2069415939
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (4.12s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25836

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9214737?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9214737?buildTab=artifacts#/) on master @ [4913247d084501a8fe63d6d5be775643b7a8531a](https://github.com/cockroachdb/cockroach/commits/4913247d084501a8fe63d6d5be775643b7a8531a):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange150727907
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange150727907
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (5.49s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9253109?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9253109?buildTab=artifacts#/) on master @ [3e68c5dc5b62e92363e015934246d2232e089ebd](https://github.com/cockroachdb/cockroach/commits/3e68c5dc5b62e92363e015934246d2232e089ebd):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2274441193
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2274441193
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (3.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9279801?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9279801?buildTab=artifacts#/) on master @ [3f113fc1a34e5bb8922298fbbb17c0c7f7ede4ac](https://github.com/cockroachdb/cockroach/commits/3f113fc1a34e5bb8922298fbbb17c0c7f7ede4ac):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange1258406796
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange1258406796
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (6.72s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9282724?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9282724?buildTab=artifacts#/) on master @ [143b63a6a27aeb286637dd2e5abddafdf0c51874](https://github.com/cockroachdb/cockroach/commits/143b63a6a27aeb286637dd2e5abddafdf0c51874):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange1271766154
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange1271766154
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (2.10s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9305792?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9305792?buildTab=artifacts#/) on master @ [54979fbbd1f7a316e2d5e60a1af8570e7a126e9f](https://github.com/cockroachdb/cockroach/commits/54979fbbd1f7a316e2d5e60a1af8570e7a126e9f):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2684245260
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2684245260
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (2.13s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Stressed this for a while on an `arm64` AWS EC2 instance without any luck at reproducing the failure, until I increased the concurrency by using `./stress -p 32 ./kvserver_test -test.run TestMergeQueueDoesNotInterruptReplicationChange`

```
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (19.19s)
    test_log_scope.go:161: test logs captured to: /tmp/logTestMergeQueueDoesNotInterruptReplicationChange969136870
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799:
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    panic.go:522: -- test log scope end --
FAIL
```

It seems like this is an issue with the test under heavy CPU utilization rather than an actual bug, so I'll plan to put up a PR to fix the test itself.

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9352579?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9352579?buildTab=artifacts#/) on master @ [7b0fffc1954bc888cd496937879270290b5cdd69](https://github.com/cockroachdb/cockroach/commits/7b0fffc1954bc888cd496937879270290b5cdd69):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange3153643673
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange3153643673
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (6.69s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestMergeQueueDoesNotInterruptReplicationChange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9369675?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/9369675?buildTab=artifacts#/) on master @ [4f3fa73a3ae1e81f8978b8548459e307d9eb73bc](https://github.com/cockroachdb/cockroach/commits/4f3fa73a3ae1e81f8978b8548459e307d9eb73bc):


```
=== RUN   TestMergeQueueDoesNotInterruptReplicationChange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2154355768
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1799: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1799
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:71
        	            				github.com/cockroachdb/cockroach/pkg/util/retry/retry.go:213
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:77
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:59
        	            				github.com/cockroachdb/cockroach/pkg/testutils/soon.go:41
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/replica_learner_test.go:1790
        	Error:      	Should be true
        	Test:       	TestMergeQueueDoesNotInterruptReplicationChange
    soon.go:71: encountered retriable replication change error: change replicas of r63 failed: descriptor changed: [expected] r63:{/Table/Max-\xfa\x00} [(n1,s1):1, next=2, gen=2, sticky=9223372036.854775807,2147483647] != [actual] r63:/{Table/Max-Max} [(n1,s1):1, next=2, gen=3, sticky=9223372036.854775807,2147483647]
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestMergeQueueDoesNotInterruptReplicationChange2154355768
--- FAIL: TestMergeQueueDoesNotInterruptReplicationChange (4.66s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueueDoesNotInterruptReplicationChange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -1732,7 +1732,9 @@ func TestMergeQueueDoesNotInterruptReplicationChange(t *testing.T) {
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 	var activateSnapshotTestingKnob int64
+	var snapshotStarted int64
 	blockSnapshot := make(chan struct{})
+	snapshotInProgress := make(chan struct{})
 	tc := testcluster.StartTestCluster(
 		t, 2, base.TestClusterArgs{
 			ReplicationMode: base.ReplicationManual,
@@ -1744,6 +1746,14 @@ func TestMergeQueueDoesNotInterruptReplicationChange(t *testing.T) {
 						DisableLoadBasedSplitting: true,
 						ReceiveSnapshot: func(_ *kvserverpb.SnapshotRequest_Header) error {
 							if atomic.LoadInt64(&activateSnapshotTestingKnob) == 1 {
+								// While the snapshot RPC should only happen once given
+								// that the cluster is running under manual replication,
+								// retries or other mechanisms can cause this to be called
+								// multiple times, so let's ensure we only close the channel
+								// snapshotInProgress once by using the snapshotStarted flag.
+								if atomic.CompareAndSwapInt64(&snapshotStarted, 0, 1) {
+									close(snapshotInProgress)
+								}
 								<-blockSnapshot
 							}
 							return nil
@@ -1772,7 +1782,7 @@ func TestMergeQueueDoesNotInterruptReplicationChange(t *testing.T) {
 	require.NoError(t, err)
 
 	select {
-	case <-time.After(100 * time.Millisecond):
+	case <-snapshotInProgress:
 	// Continue.
 	case <-replicationChange:
 		t.Fatal("did not expect the replication change to complete")
@@ -1787,18 +1797,16 @@ func TestMergeQueueDoesNotInterruptReplicationChange(t *testing.T) {
 	// TestCluster currently overrides this when used with ReplicationManual.
 	db.Exec(t, `SET CLUSTER SETTING kv.range_merge.queue_enabled = true`)
 
-	testutils.SucceedsSoon(t, func() error {
-		// While this replication change is stalled, we'll trigger a merge and
-		// ensure that the merge correctly notices that there is a snapshot in
-		// flight and ignores the range.
-		store, repl := getFirstStoreReplica(t, tc.Server(0), scratchKey)
-		_, processErr, enqueueErr := store.Enqueue(
-			ctx, "merge", repl, true /* skipShouldQueue */, false, /* async */
-		)
-		require.NoError(t, enqueueErr)
-		require.True(t, kvserver.IsReplicationChangeInProgressError(processErr))
-		return nil
-	})
+	// While this replication change is stalled, we'll trigger a merge and
+	// ensure that the merge correctly notices that there is a snapshot in
+	// flight and ignores the range.
+	store, repl := getFirstStoreReplica(t, tc.Server(0), scratchKey)
+	_, processErr, enqueueErr := store.Enqueue(
+		ctx, "merge", repl, true /* skipShouldQueue */, false, /* async */
+	)
+	require.NoError(t, enqueueErr)
+	require.Truef(t, kvserver.IsReplicationChangeInProgressError(processErr),
+		"expected replication change in progress error, got %+v", processErr)
 }
 
 func TestMergeQueueSeesLearnerOrJointConfig(t *testing.T) {

